index.js

```diff
@@ -4,15 +4,135 @@
 
   'use strict';
 
-  return function(){};
+  var $alt = 'fantasy-land/alt';
+  var $ap = 'fantasy-land/ap';
+  var $map = 'fantasy-land/map';
+  var $of = 'fantasy-land/of';
+  var $zero = 'fantasy-land/zero';
+  var $$type = '@@type';
+
+  var ordinal = ['first', 'second', 'third', 'fourth', 'fifth'];
+  var show = Z.toString;
+
+  function isFunction(f){
+    return typeof f === 'function';
+  }
+
+  function isBinary(f){
+    return f.length >= 2;
+  }
+
+  function isApplicativeRepr(Repr){
+    return isFunction(Repr[$of]) && Z.Applicative.test(Repr[$of]());
+  }
+
+  function invalidArgument(it, at, expected, actual){
+    throw new TypeError(
+      it
+      + ' expects its '
+      + ordinal[at]
+      + ' argument to '
+      + expected
+      + '\n  Actual: '
+      + show(actual)
+    );
+  }
+
+  function invalidContext(it, actual){
+    throw new TypeError(
+      it + ' was invoked outside the context of a Concurrently. \n  Called on: ' + show(actual)
+    );
+  }
+
+  function getReprType(Repr){
+    return Repr[$$type] || Repr.name || 'Anonymous';
+  }
+
+  function concurrentPrepender(x, i, xs){
+    return i === xs.length - 1 ? ('Concurrent' + x) : x;
+  }
+
+  function computeType(type){
+    return type.split('/').map(concurrentPrepender).join('/');
+  }
+
+  //concurrify :: Applicative m
+  //           => (TypeRep m, () -> m a, (m a, m a) -> m a, (m a, m (a -> b)) -> m b)
+  //           -> Concurrently m
+  return function concurrify(Repr, zero, race, ap){
+
+    if(!isApplicativeRepr(Repr)) invalidArgument('concurrify', 0, 'represent an Applicative', Repr);
+    if(!isFunction(zero)) invalidArgument('concurrify', 1, 'be a function', zero);
+    if(!isFunction(race)) invalidArgument('concurrify', 2, 'be a function', race);
+    if(!isBinary(race)) invalidArgument('concurrify', 2, 'be binary', race);
+    if(!isFunction(ap)) invalidArgument('concurrify', 3, 'be a function', ap);
+    if(!isBinary(ap)) invalidArgument('concurrify', 3, 'be binary', ap);
+
+    var INNERTYPE = getReprType(Repr);
+    var OUTERTYPE = computeType(INNERTYPE);
+
+    function Concurrently(sequential){
+      this.sequential = sequential;
+    }
+
+    function isInner(x){
+      return x instanceof Repr
+      || (Boolean(x) && x.constructor === Repr)
+      || type(x) === Repr[$$type];
+    }
+
+    function isOuter(x){
+      return x instanceof Concurrently
+      || (Boolean(x) && x.constructor === Concurrently)
+      || type(x) === OUTERTYPE;
+    }
+
+    function construct(x){
+      if(!isInner(x)) invalidArgument(OUTERTYPE, 0, 'be of type "' + INNERTYPE + '"', type(x));
+      return new Concurrently(x);
+    }
+
+    var proto = Concurrently.prototype = construct.prototype = {
+      sequential: null,
+      constructor: construct
+    };
+
+    construct[$$type] = OUTERTYPE;
+
+    construct[$zero] = zero;
+
+    construct[$of] = function Concurrently$of(value){
+      return new Concurrently(Repr[$of](value));
```

Comment:
I think yes, as theoretically Promise could be "fixed" and sanctuary-type-classes could dispatches to it's apropriate methods